### ISSUE TITLE
Implementation is only fast because it's on "steroids"

### ISSUE THREAD

**Issue by DaOnlyOwner** [lnishan/vector#7]
Hi,

**Comment by lnishan** [comment]
Hello,
Thanks for the comment.
This is mostly just a toy project for me.

1. It's interesting how you referenced the Facebook write-up because it is flawed. That argument assumes that you will always have your new memory segment allocated **right next to** the previous one, which is obviously incorrect. By removing this false assumption, it's clear that it's still very possible that we get to reuse our old memory segments.  
Also, C++11 added the capability for programmers to inspect and control the **capacity** of the current vector. Since this implementation is mostly C++-17 compliant, this option is available to you. Is a growth factor of 4 significant? Yes, but it doesn't really matter while you're inserting the initial elements of a large sequence. You can call `shrink_to_fit()` after inserting the elements and even insert run-time checks to prevent memory from peaking up. A growth factor of 2 or even less just doesn't work optimally in all use cases.

2. This is also false. It's widely-known that standard implementations are oftentimes less than optimal. (See [EASTL](https://github.com/electronicarts/EASTL)). Both `unordered_map` and `unordered_set` especially are both known to be incredibly slow. STLs are usually implemented for generic use, so if you have specialized uses or your input sequence exhibits a particular characteristic, you should never use STLs if performance is critical.

**Comment by DaOnlyOwner** [comment]
Hi,
I get that ;) It's still a nice project.

1. Let's say we have a memory block allocated of size 2^n. (n is the number of times we had to resize)
    Then we have to expand the vector again. With a factor == 2 this yields n_new =  2^(n+1); Can our 
    malloc() implementation reuse old memory blocks to store the n_new block? No it cannot because
                         `1 (first resize) + 2 (second resize) + 4 + 8 + 16 + ... 2^(n-1) = (2^n)-1. `
    Note that we cannot free the 2^n block yet. So our vector has to expand forward. (See 
    https://crntaylor.wordpress.com/2011/07/15/optimal-memory-reallocation-and-the-golden-ratio/ 
    maybe). But you are correct if you say that the argument in the article is false. It doesn't provide better 
    performance but better memory reuse. I would suggest to use the factor 1.5 and initially allocate 
    memory on the stack until you have to put it on the heap.  

2.  The motivation for EASTL (see http://www.open-
     std.org/jtc1/sc22/wg21/docs/papers/2007/n2271.html) has some valid points, for example the 
     allocator design, but some of them are simply not true anymore, For example: "Existing std STL  
     implementations use deep function calls. This results in low performance with compilers that are 
     weak at inlining, as is the currently prevalent open-source C++ compiler. See Appendix item 15 and 
     Appendix item 10". But overall this is correct of course.

**Comment by dufferzafar** [comment]
I'm a beginner C++ programmer and I loved reading your discussion, even though I only understood like 30% of it.

@DaOnlyOwner On your quest, what did you finally find? Like if I need to write performance critical code, should I look for such alternative implementations or just stick with the STL that my compiler provides?

**Comment by dufferzafar** [comment]
Thanks for your detailed response. All this is very informative.

> Did I mention the excellent iteration performance of std::vector? (Don't use std::list etc.)

This is due to locality of reference of caches, right?

**Comment by DaOnlyOwner** [comment]
Yes, this is because the vector is implemented as a contiguous array.

**Comment by dufferzafar** [comment]
Yeah, there's a video by Stroustrup: [Why you should avoid linked lists?](https://www.youtube.com/watch?v=YQs6IC-vgmo)
